src/pypromice/L0toL1.py

```diff
@@ -177,6 +191,206 @@ def _flagNAN(ds, flag_file=None):
+                tmp[tmp > val] = np.nan
+                ds_out[var].loc[dict(time=slice(t0, t1))] = tmp
+                
+            if func == "upper_perc_filter":
+                tmp = ds_out[var].loc[dict(time=slice(t0, t1))].copy()
+                df_w = ds_out[var].loc[dict(time=slice(t0, t1))].resample("14D").quantile(1 - val / 100)
+                df_w = ds_out[var].loc[dict(time=slice(t0, t1))].resample("14D").var()
+                for m_start, m_end in zip(df_w.time[:-2], df_w.time[1:]):
+                    msk = (tmp.time >= m_start) & (tmp.time < m_end)
+                    values_month = tmp.loc[msk].values
+                    values_month[values_month < df_w.loc[m_start]] = np.nan
+                    tmp.loc[msk] = values_month
+
+                ds_out[var].loc[dict(time=slice(t0, t1))] = tmp.values
+
+            if func == "biweekly_upper_range_filter":
+                tmp = ds_out[var].loc[dict(time=slice(t0, t1))].copy()
+                df_max = ds_out[var].loc[dict(time=slice(t0, t1))].resample("14D").max()
+                for m_start, m_end in zip(df_max.time[:-2], df_max.time[1:]):
+                    msk = (tmp.time >= m_start) & (tmp.time < m_end)
+                    lim = df_max.loc[m_start] - val
+                    values_month = tmp.loc[msk].values
+                    values_month[values_month < lim] = np.nan
+                    tmp.loc[msk] = values_month
+                # remaining samples following outside of the last 2 weeks window
+                msk = tmp.time >= m_end
+                lim = df_max.loc[m_start] - val
+                values_month = tmp.loc[msk].values
+                values_month[values_month < lim] = np.nan
+                tmp.loc[msk] = values_month
+                # updating original pandas
+                ds_out[var].loc[dict(time=slice(t0, t1))] = tmp.values
+
+            if func == "hampel_filter":
+                tmp = ds_out[var].loc[dict(time=slice(t0, t1))]
+                tmp = _hampel(tmp, k=7 * 24, t0=val)
+                ds_out[var].loc[dict(time=slice(t0, t1))] = tmp.values
+
+            if func == "grad_filter":
+                tmp = ds_out[var].loc[dict(time=slice(t0, t1))].copy()
+                msk = ds_out[var].loc[dict(time=slice(t0, t1))].copy().diff()
+                tmp[np.roll(msk.abs() > val, -1)] = np.nan
+                ds_out[var].loc[dict(time=slice(t0, t1))] = tmp
+
+            if "swap_with_" in func:
+                var2 = func[10:]
+                val_var = ds_out[var].loc[dict(time=slice(t0, t1))].values.copy()
+                val_var2 = ds_out[var2].loc[dict(time=slice(t0, t1))].values.copy()
+                ds_out[var2].loc[dict(time=slice(t0, t1))] = val_var
+                ds_out[var].loc[dict(time=slice(t0, t1))] = val_var2
+
+            if func == "rotate":
+                ds_out[var].loc[dict(time=slice(t0, t1))] = (ds_out[var].loc[dict(time=slice(t0, t1))].values + val) % 360
```

Comment:
This looks great generally. I'd like to test it's robustness, but looks good and thorough at first glance.